docs/rely/mock-functions.md

```diff
@@ -0,0 +1,209 @@
+---
+id: mock-functions
+title: Mock Functions
+sidebar_label: Mock Functions
+---
+
+Mock functions (or "spies") in Rely can be used to track information about the arguments and return values of functions and allow for test-time configuration of return values. Rely has built in matchers for dealing with mock functions that are documented extensively [here](expect.md#expectmock).
+
+> Unlike Jest in JavaScript, module level mocking is not supported. The functionality available in Rely is analagous to what Jest offers via [`jest.fn` ](https://jestjs.io/docs/en/jest-object#jestfnimplementation) and there is no direct [`jest.mock`](https://jestjs.io/docs/en/jest-object#jestmockmodulename-factory-options) analogy. Many situations in which `jest.mock` would be useful can be addressed through dependency management and the structuring of side effects within your application. A simple example using functors is shown [below](#injecting-via-functor).
+
+## Using Mock Functions
+
+### Creating Mocks
+
+There are constructors for creating mocks that take up to 7 arguments. A default implementation must be specified.
+
+```reason
+open TestFramework;
+
+let calculateInterest = Mock.mock3((p, r, t) => p *. exp(r *. t));
+```
+
+```reason
+module Mock: {
+  type t('fn, 'returnType, 'tupledArgs);
+
+  let mock1: ('args => 'ret) => t('args => 'ret, 'ret, 'args);
+
+  let mock2:
+    (('arg1, 'arg2) => 'ret) =>
+    t(('arg1, 'arg2) => 'ret, 'ret, ('arg1, 'arg2));
+
+  let mock3:
+    (('arg1, 'arg2, 'arg3) => 'ret) =>
+    t(('arg1, 'arg2, 'arg3) => 'ret, 'ret, ('arg1, 'arg2, 'arg3));
+
+  let mock4:
+    (('arg1, 'arg2, 'arg3, 'arg4) => 'ret) =>
+    t(
+      ('arg1, 'arg2, 'arg3, 'arg4) => 'ret,
+      'ret,
+      ('arg1, 'arg2, 'arg3, 'arg4),
+    );
+
+  let mock5:
+    (('arg1, 'arg2, 'arg3, 'arg4, 'arg5) => 'ret) =>
+    t(
+      ('arg1, 'arg2, 'arg3, 'arg4, 'arg5) => 'ret,
+      'ret,
+      ('arg1, 'arg2, 'arg3, 'arg4, 'arg5),
+    );
+
+  let mock6:
+    (('arg1, 'arg2, 'arg3, 'arg4, 'arg5, 'arg6) => 'ret) =>
+    t(
+      ('arg1, 'arg2, 'arg3, 'arg4, 'arg5, 'arg6) => 'ret,
+      'ret,
+      ('arg1, 'arg2, 'arg3, 'arg4, 'arg5, 'arg6),
+    );
+
+  let mock7:
+    (('arg1, 'arg2, 'arg3, 'arg4, 'arg5, 'arg6, 'arg7) => 'ret) =>
+    t(
+      ('arg1, 'arg2, 'arg3, 'arg4, 'arg5, 'arg6, 'arg7) => 'ret,
+      'ret,
+      ('arg1, 'arg2, 'arg3, 'arg4, 'arg5, 'arg6, 'arg7),
+    );
+};
+```
+
+### Calling Mock Functions
+
+The `Mock.mockN` constructor returns a mutable opaque type for tracking call information. To access the underlying function, use `Mock.fn`.
+
+```reason
+open TestFramework;
+
+describe("Mock documentation", ({test}) => {
+  test("calling a mock function", ({expect}) => {
+    let mockDouble = Mock.mock1(x => x * 2);
+    let mockFunction = Mock.fn(mockDouble);
+
+    expect.int(mockFunction(2)).toBe(4);
+  })
```

Comment:
;

Reply:
The formatter actually removes it so I'll leave this without a `'`